astropy/coordinates/representation/base.py

```diff
@@ -143,10 +145,26 @@ class BaseRepresentationOrDifferential(MaskableShapedLikeNDArray):
 
     # Ensure multiplication/division with ndarray or Quantity doesn't lead to
     # object arrays.
-    __array_priority__ = 50000
+    __array_priority__: ClassVar[int] = 50000
+
+    name: ClassVar[str]
+    """Name of the representation or differential.
+
+    In lower case, with any trailing 'representation' or 'differential' removed. (E.g.,
+    'spherical' for `~astropy.coordinates.SphericalRepresentation` or
+    `~astropy.coordinates.SphericalDifferential`.)
+    """
 
     info = BaseRepresentationOrDifferentialInfo()
 
+    def __init_subclass__(cls) -> None:
+        # Name of the representation or differential.
+        cls.name = (
```

Comment:
In frames, we are a bit more careful: we only set the name if `cls.name` does not already exist *or* if `cls.name` is inherited from a superclass, so should be overwritten. Like in registration, we probably should also not give any `Base*` class a name.

Reply:
Not registering is fine, but not setting an attribute is dangerous!
Every class should have a name, otherwise you can get an error when doing `getattr(cls, "name")` even though the annotations promise that the API has the name attribute.

Reply:
I guess the underlying issue is that our `Base*` classes are really more like abstract classes, so I guess `name` is more like an `abstractattribue`. Anyway, for now fine to just make sure `name` is what `get_name()` would have returned!

Reply:
Unresolving just to deal with the case where a class defines its own name - I don't think we should override that. But that means also doing a check like that in `BaseCoordinateFrame` to ensure it is not just inherited.

Obviously, since this was not used so far, we can leave it as is. That said, it was possible before, by overriding `get_name()`...

Reply:
Fixed!